Attachments Pull Test Failure

Slightly different failure this time:

```
com.cloudant.sync.replication.AttachmentsPullTest > pullRevisionsWithBigAttachments[0] FAILED
    java.lang.AssertionError
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at com.cloudant.sync.replication.AttachmentsPullTest.pull(AttachmentsPullTest.java:318)
        at com.cloudant.sync.replication.AttachmentsPullTest.pullRevisionsWithBigAttachments(AttachmentsPullTest.java:103)
```

Although I merged that extra debug it turns out that some of the other tests decide to catch the exceptions and hide the causes e.g.:

>com.cloudant.sync.replication.AttachmentsPullTest > pullRevisionsWithBigAttachments[0] FAILED
    java.lang.AssertionError: Create/pull error java.lang.Exception: Replication errored
        at org.junit.Assert.fail(Assert.java:88)
        at com.cloudant.sync.replication.AttachmentsPullTest.pullRevisionsWithBigAttachments(AttachmentsPullTest.java:105)

This means we'll need some more test changes to try and expose the real errors.

Another failure mode:

```
com.cloudant.sync.replication.AttachmentsPullTest > pullRevisionsWithBigAttachments[0] FAILED
    java.lang.AssertionError: Attachment is null
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertNotNull(Assert.java:621)
        at com.cloudant.sync.replication.AttachmentsPullTest.pullRevisionsWithBigAttachments(AttachmentsPullTest.java:108)
```

--- a/AndroidTest/app/build.gradle
+++ b/AndroidTest/app/build.gradle
@@ -69,7 +69,7 @@ android {
     }
 
     adbOptions {
-        timeOutInMs 3*60*1000 // 3 minutes in ms.
+        timeOutInMs 5*60*1000 // 5 minutes in ms.
     }
 
 }

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,3 +1,6 @@
+# Unreleased
+- [IMPROVED] Increased the resilience of replication to network failures.
+
 # 2.0.0 (2017-02-07)
 - [BREAKING CHANGE] With the release of version 2.0 of the library,
   there are a large number of breaking changes to package names,

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/mazha/CouchClient.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/mazha/CouchClient.java
@@ -21,17 +21,16 @@
 package com.cloudant.sync.internal.mazha;
 
 
-import com.cloudant.sync.internal.common.RetriableTask;
 import com.cloudant.http.Http;
 import com.cloudant.http.HttpConnection;
 import com.cloudant.http.HttpConnectionRequestInterceptor;
 import com.cloudant.http.HttpConnectionResponseInterceptor;
+import com.cloudant.sync.internal.common.RetriableTask;
 import com.cloudant.sync.internal.documentstore.DocumentRevsList;
 import com.cloudant.sync.internal.documentstore.MultipartAttachmentWriter;
 import com.cloudant.sync.internal.util.JSONUtils;
 import com.cloudant.sync.internal.util.Misc;
 import com.fasterxml.jackson.core.type.TypeReference;
-import com.fasterxml.jackson.databind.JavaType;
 
 import org.apache.commons.io.IOUtils;
 
@@ -49,6 +48,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.Callable;
+import java.util.logging.Level;
 import java.util.logging.Logger;
 
 public class CouchClient  {
@@ -200,21 +200,38 @@ public class CouchClient  {
 
     // execute HTTP task with retries:
     // return an InputStream if successful or throw an exception
-    private InputStream executeToInputStreamWithRetry(final Callable<ExecuteResult> task) throws CouchException {
+    private <T> T executeWithRetry(final Callable<ExecuteResult> task,
+                                   InputStreamProcessor<T> processor) throws
+            CouchException {
         int attempts = 10;
-        CouchException lastException= null;
+        CouchException lastException = null;
         while (attempts-- > 0) {
             ExecuteResult result = null;
             try {
                 result = task.call();
+                if (result.stream != null) {
+                    // success - process the inputstream
+                    try {
+                        return processor.processStream(result.stream);
+                    } catch (Exception e) {
+                        // Error processing the input stream
+                        if (attempts > 0) {
+                            logger.log(Level.WARNING, "Received an exception during response " +
+                                    "stream processing. A retry will be attempted.", e);
+                        } else {
+                            logger.log(Level.SEVERE, "Response stream processing failed, no " +
+                                    "retries remaining.", e);
+                            throw e;
+                        }
+                    } finally {
+                        result.stream.close();
+                    }
+                }
             } catch (Exception e) {
                 throw new CouchException("Unexpected exception", e, -1);
             }
             lastException = result.exception;
-            if (result.stream != null) {
-                // success - return the inputstream
-                return result.stream;
-            } else if (result.fatal) {
+            if (result.fatal) {
                 // fatal exception - don't attempt any more retries
                 throw result.exception;
             }
@@ -222,30 +239,29 @@ public class CouchClient  {
         throw lastException;
     }
 
-    private <T> T executeToJsonObjectWithRetry(final HttpConnection connection,
-                                               Class<T> c) throws CouchException {
-        InputStream is = this.executeToInputStreamWithRetry(connection);
-        InputStreamReader isr = new InputStreamReader(is, Charset.forName("UTF-8"));
-        try {
-            T json = JSONUtils.fromJson(isr, c);
-            return json;
-        } finally {
-            IOUtils.closeQuietly(is);
-        }
+    private <T> T executeToJsonObjectWithRetry(final HttpConnection connection, final
+    TypeReference<T> type) throws CouchException {
+        return executeWithRetry(connection, new TypeInputStreamProcessor<T>(type));
     }
 
-    private InputStream executeToInputStreamWithRetry(final HttpConnection connection) throws CouchException {
+    private <T> T executeToJsonObjectWithRetry(final HttpConnection connection,
+                                               Class<T> c) throws CouchException {
+        return executeToJsonObjectWithRetry(connection, new CouchClientTypeReference<T>(c));
+    }
+
+    private <T> T executeWithRetry(final HttpConnection connection,
+                                                InputStreamProcessor<T> processor) throws
+            CouchException {
         // all CouchClient requests want to receive application/json responses
         connection.requestProperties.put("Accept", "application/json");
         connection.responseInterceptors.addAll(responseInterceptors);
         connection.requestInterceptors.addAll(requestInterceptors);
-        InputStream is = this.executeToInputStreamWithRetry(new Callable<ExecuteResult>() {
+        return this.executeWithRetry(new Callable<ExecuteResult>() {
             @Override
             public ExecuteResult call() throws Exception {
                 return execute(connection);
             }
-        });
-        return is;
+        }, processor);
     }
 
     public void createDb() {
@@ -307,7 +323,7 @@ public class CouchClient  {
         URI doc = this.uriHelper.documentUri(id);
         try {
             HttpConnection connection = Http.HEAD(doc);
-            this.executeToInputStreamWithRetry(connection);
+            this.executeWithRetry(connection, new NoOpInputStreamProcessor());
             return true;
         } catch (Exception e) {
             return false;
@@ -335,17 +351,7 @@ public class CouchClient  {
         }
     }
 
-    public InputStream getDocumentStream(String id, String rev) {
-        Misc.checkNotNullOrEmpty(id, "id");
-        Misc.checkNotNullOrEmpty(rev, "rev");
-        Map<String, Object> queries = new HashMap<String, Object>();
-        queries.put("rev", rev);
-        final URI doc = this.uriHelper.documentUri(id, queries);
-        HttpConnection connection = Http.GET(doc);
-        return executeToInputStreamWithRetry(connection);
-    }
-
-    public InputStream getAttachmentStream(String id, String rev, String attachmentName, final boolean acceptGzip) {
+    public <T> T processAttachmentStream(String id, String rev, String attachmentName, final boolean acceptGzip, InputStreamProcessor<T> processor) {
         Misc.checkNotNullOrEmpty(id, "id");
         Misc.checkNotNullOrEmpty(rev, "rev");
         Map<String, Object> queries = new HashMap<String, Object>();
@@ -355,7 +361,7 @@ public class CouchClient  {
         if (acceptGzip) {
             connection.requestProperties.put("Accept-Encoding", "gzip");
         }
-        return executeToInputStreamWithRetry(connection);
+        return executeWithRetry(connection, processor);
     }
 
     public void putAttachmentStream(String id, String rev, String attachmentName, String contentType, byte[] attachmentData) {
@@ -366,7 +372,8 @@ public class CouchClient  {
         URI doc = this.uriHelper.attachmentUri(id, queries, attachmentName);
         HttpConnection connection = Http.PUT(doc, contentType);
         connection.setRequestBody(attachmentData);
-        this.executeToInputStreamWithRetry(connection);
+        //TODO check result?
+        executeToJsonObjectWithRetry(connection, Response.class);
     }
 
     /**
@@ -385,7 +392,7 @@ public class CouchClient  {
     public Map<String,Object> getDocConflictRevs(String id)  {
         Map<String, Object> options = new HashMap<String, Object>();
         options.put("conflicts", true);
-        return this.getDocument(id, options, JSONUtils.mapStringToObject());
+        return this.getDocument(id, options, JSONUtils.STRING_MAP_TYPE_DEF);
     }
 
     /**
@@ -417,7 +424,7 @@ public class CouchClient  {
             options.put("att_encoding_info", true);
         }
         options.put("open_revs", JSONUtils.toJson(revisions));
-        return this.getDocument(id, options, JSONUtils.openRevisionList());
+        return this.getDocument(id, options, JSONUtils.OPEN_REVS_LIST_TYPE_DEF);
     }
 
     /**
@@ -478,30 +485,24 @@ public class CouchClient  {
     }
 
     public Map<String, Object> getDocument(String id) {
-        return this.getDocument(id, new HashMap<String, Object>(), JSONUtils.mapStringToObject());
+        return this.getDocument(id, new HashMap<String, Object>(), JSONUtils.STRING_MAP_TYPE_DEF);
     }
 
     public <T> T getDocument(String id, final Class<T> type)  {
-        JavaType javaType = JSONUtils.getTypeFactory().constructType(type);
-        return this.getDocument(id, new HashMap<String, Object>(), javaType);
+        return this.getDocument(id, new HashMap<String, Object>(),
+                new CouchClientTypeReference<T>(type));
     }
 
-    public <T> T getDocument(final String id, final Map<String, Object> options, final JavaType type) {
+    public <T> T getDocument(final String id, final Map<String, Object> options,
+                             final TypeReference<T> type) {
         Misc.checkNotNullOrEmpty(id, "id");
         Misc.checkNotNull(type, "Type");
 
         URI doc = uriHelper.documentUri(id, options);
-        InputStream is = null;
-        try {
-            HttpConnection connection = Http.GET(doc);
-            is = executeToInputStreamWithRetry(connection);
-            T returndoc = JSONUtils.fromJson(new InputStreamReader(is, Charset.forName("UTF-8"))
-                    , type);
-            logger.fine("get returning " + returndoc);
-            return returndoc;
-        } finally {
-            closeQuietly(is);
-        }
+        HttpConnection connection = Http.GET(doc);
+        T returndoc = executeToJsonObjectWithRetry(connection, type);
+        logger.fine("get returning " + returndoc);
+        return returndoc;
     }
 
     public Map<String, Object> getDocument(String id, String rev) {
@@ -513,16 +514,14 @@ public class CouchClient  {
         Misc.checkNotNullOrEmpty(rev, "rev");
         Misc.checkNotNull(type, "Type");
 
-        InputStream is = null;
-        try {
-            is = this.getDocumentStream(id, rev);
-            T returndoc = JSONUtils.fromJson(new InputStreamReader(is, Charset.forName("UTF-8"))
-                    , type);
-            logger.fine("get returning " + returndoc);
-            return returndoc;
-        } finally {
-            closeQuietly(is);
-        }
+        Map<String, Object> queries = new HashMap<String, Object>();
+        queries.put("rev", rev);
+        final URI doc = this.uriHelper.documentUri(id, queries);
+        HttpConnection connection = Http.GET(doc);
+        T returndoc = executeToJsonObjectWithRetry(connection, type);
+        logger.fine("get returning " + returndoc);
+        return returndoc;
+
     }
 
     public <T> T getDocument(String id, String rev, final Class<T> type) {
@@ -536,25 +535,19 @@ public class CouchClient  {
      *
      */
     public DocumentRevs getDocRevisions(String id, String rev) {
-        return getDocRevisions(id, rev, JSONUtils.documentRevs());
+        return getDocRevisions(id, rev,
+                new CouchClientTypeReference<DocumentRevs>(DocumentRevs.class));
     }
 
-    public <T> T getDocRevisions(String id, String rev, JavaType type) {
+    public <T> T getDocRevisions(String id, String rev, TypeReference<T> type) {
         Misc.checkNotNullOrEmpty(id, "id");
         Misc.checkNotNull(rev, "Revision ID");
         Map<String, Object> queries = new HashMap<String, Object>();
         queries.put("revs", "true");
         queries.put("rev", rev);
         URI findRevs = this.uriHelper.documentUri(id, queries);
-
-        InputStream is = null;
-        try {
-            HttpConnection connection = Http.GET(findRevs);
-            is = this.executeToInputStreamWithRetry(connection);
-            return JSONUtils.fromJson(new InputStreamReader(is, Charset.forName("UTF-8")), type);
-        } finally {
-            closeQuietly(is);
-        }
+        HttpConnection connection = Http.GET(findRevs);
+        return executeToJsonObjectWithRetry(connection, type);
     }
 
     // Document should be complete document include "_id" matches ID
@@ -583,40 +576,23 @@ public class CouchClient  {
         return executeToJsonObjectWithRetry(connection, Response.class);
     }
 
-    private void closeQuietly(InputStream is) {
-        try {
-            if (is != null) {
-                is.close();
-            }
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
-    }
-
-    private InputStream bulkCreateDocsInputStream(List<?> objects) {
-        Misc.checkNotNull(objects, "Object list");
-        String newEditsVal = "\"new_edits\": false, ";
-        URI uri = this.uriHelper.bulkDocsUri();
-        String payload = String.format("{%s%s%s}", newEditsVal, "\"docs\": ",
-                JSONUtils.toJson(objects));
-        HttpConnection connection = Http.POST(uri, "application/json");
-        connection.setRequestBody(payload);
-        return this.executeToInputStreamWithRetry(connection);
-    }
-
     public List<Response> bulkCreateDocs(Object... objects) {
         return bulkCreateDocs(Arrays.asList(objects));
     }
 
     public List<Response> bulkCreateDocs(List<?> objects) {
-        InputStream is = null;
-        try {
-            is = bulkCreateDocsInputStream(objects);
-            return JSONUtils.fromJsonToList(new InputStreamReader(is, Charset.forName("UTF-8")),
-                    new CouchClientTypeReference<List<Response>>());
-        } finally {
-            closeQuietly(is);
-        }
+        Misc.checkNotNull(objects, "Object list");
+        String newEditsVal = "\"new_edits\": false, ";
+        String payload = String.format("{%s%s%s}", newEditsVal, "\"docs\": ",
+                JSONUtils.toJson(objects));
+        return bulkCreateDocs(payload);
+    }
+
+    private List<Response> bulkCreateDocs(String payload) {
+        URI uri = this.uriHelper.bulkDocsUri();
+        HttpConnection connection = Http.POST(uri, "application/json");
+        connection.setRequestBody(payload);
+        return executeToJsonObjectWithRetry(connection, new CouchClientTypeReference<List<Response>>());
     }
 
     /**
@@ -641,17 +617,7 @@ public class CouchClient  {
     public List<Response> bulkCreateSerializedDocs(List<String> serializedDocs) {
         Misc.checkNotNull(serializedDocs, "Serialized doc list");
         String payload = generateBulkSerializedDocsPayload(serializedDocs);
-        URI uri = this.uriHelper.bulkDocsUri();
-        InputStream is = null;
-        HttpConnection connection = Http.POST(uri, "application/json");
-        connection.setRequestBody(payload);
-        try {
-            is = this.executeToInputStreamWithRetry(connection);
-            return JSONUtils.fromJsonToList(new InputStreamReader(is, Charset.forName("UTF-8")),
-                    new CouchClientTypeReference<List<Response>>());
-        } finally {
-            closeQuietly(is);
-        }
+        return bulkCreateDocs(payload);
     }
 
     private String generateBulkSerializedDocsPayload(List<String> serializedDocs) {
@@ -692,21 +658,13 @@ public class CouchClient  {
      * @see <a target="_blank" href="http://wiki.apache.org/couchdb/HttpPostRevsDiff">HttpPostRevsDiff documentation</a>
      */
     public Map<String, MissingRevisions> revsDiff(Map<String, Set<String>> revisions) {
-        Misc.checkNotNull(revisions,"Input revisions");
+        Misc.checkNotNull(revisions, "Input revisions");
         URI uri = this.uriHelper.revsDiffUri();
         String payload = JSONUtils.toJson(revisions);
-        InputStream is = null;
-        try {
-            HttpConnection connection = Http.POST(uri, "application/json");
-            connection.setRequestBody(payload);
-            is = executeToInputStreamWithRetry(connection);
 
-
-            return JSONUtils.fromJson(new InputStreamReader(is, Charset.forName("UTF-8")),
-                    JSONUtils.mapStringMissingRevisions());
-        } finally {
-            closeQuietly(is);
-        }
+        HttpConnection connection = Http.POST(uri, "application/json");
+        connection.setRequestBody(payload);
+        return executeToJsonObjectWithRetry(connection, JSONUtils.STRING_MISSING_REVS_MAP_TYPE_DEF);
     }
 
     public Response putMultipart(final MultipartAttachmentWriter mpw) {
@@ -742,6 +700,9 @@ public class CouchClient  {
         public Set<String> missing;
     }
 
+    public interface InputStreamProcessor<T> {
+        T processStream(InputStream stream) throws Exception;
+    }
 
     private static class CouchClientTypeReference<T> extends TypeReference<T> {
 
@@ -764,4 +725,29 @@ public class CouchClient  {
 
         }
     }
-}
+
+    private static final class TypeInputStreamProcessor<T> implements InputStreamProcessor<T> {
+
+        private final TypeReference<T> typeReference;
+        TypeInputStreamProcessor(TypeReference<T> typeReference) {
+            this.typeReference = typeReference;
+        }
+
+        @Override
+        public T processStream(InputStream stream) {
+            return JSONUtils.fromJson(new InputStreamReader(stream, Charset.forName
+                    ("UTF-8")), typeReference);
+        }
+    }
+
+    /**
+     * For use with e.g. HEAD requests where there is no InputStream
+     */
+    public static final class NoOpInputStreamProcessor implements InputStreamProcessor<Void> {
+
+        @Override
+        public Void processStream(InputStream stream) {
+            return null;
+        }
+    }
+ }

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/mazha/CouchClient.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/mazha/CouchClient.java
@@ -320,10 +320,8 @@ public class CouchClient  {
     // TODO does this still work the same way we expect it to?
     public boolean contains(String id) {
         Misc.checkNotNullOrEmpty(id, "id");
-        URI doc = this.uriHelper.documentUri(id);
         try {
-            HttpConnection connection = Http.HEAD(doc);
-            this.executeWithRetry(connection, new NoOpInputStreamProcessor());
+            getDocumentRev(id);
             return true;
         } catch (Exception e) {
             return false;
@@ -554,16 +552,34 @@ public class CouchClient  {
     public Response update(String id, Object document) {
         Misc.checkNotNullOrEmpty(id, "id");
         Misc.checkNotNull(document, "Document");
-        if (!this.contains(id)) {
-            throw new NoResourceException("No document for given id: " + id);
-        }
 
-        String json = JSONUtils.toJson(document);
+        // Get the latest rev, will throw if doc doesn't exist
+        getDocumentRev(id);
+
+        return putUpdate(id, document);
+    }
+
+    public Response putUpdate(String id, Object document) {
+        Misc.checkNotNullOrEmpty(id, "id");
+        Misc.checkNotNull(document, "Document");
+
         URI doc = this.uriHelper.documentUri(id);
+        String json = JSONUtils.toJson(document);
         HttpConnection connection = Http.PUT(doc, "application/json");
         connection.setRequestBody(json);
         Response r = executeToJsonObjectWithRetry(connection, Response.class);
         return r;
+    }
+
+    public String getDocumentRev(String id) {
+        URI doc = this.uriHelper.documentUri(id);
+
+        HttpConnection head = Http.HEAD(doc);
+        executeWithRetry(head, new NoOpInputStreamProcessor());
+
+        String rev = head.getConnection().getHeaderField("ETag");
+        // Remove enclosing "" before returning
+        return rev.substring(1, rev.length()-1);
     }
 
     public Response delete(String id, String rev) {
@@ -701,6 +717,7 @@ public class CouchClient  {
     }
 
     public interface InputStreamProcessor<T> {
+
         T processStream(InputStream stream) throws Exception;
     }
 

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/AttachmentPullProcessor.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/AttachmentPullProcessor.java
@@ -1,0 +1,49 @@
+/*
+ * Copyright © 2017 IBM Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package com.cloudant.sync.internal.replication;
+
+import com.cloudant.sync.documentstore.Attachment;
+import com.cloudant.sync.documentstore.AttachmentException;
+import com.cloudant.sync.documentstore.UnsavedStreamAttachment;
+import com.cloudant.sync.internal.documentstore.PreparedAttachment;
+import com.cloudant.sync.internal.mazha.CouchClient;
+
+import java.io.InputStream;
+
+public class AttachmentPullProcessor implements CouchClient
+        .InputStreamProcessor<PreparedAttachment> {
+
+    private final DatastoreWrapper datastoreWrapper;
+    private final String contentType;
+    private final Attachment.Encoding encoding;
+    private final long length;
+    private final long encodedLength;
+
+    AttachmentPullProcessor(DatastoreWrapper wrapper, String name, String contentType, String
+            encoding, long length, long encodedLength) {
+        this.datastoreWrapper = wrapper;
+        this.contentType = contentType;
+        this.encoding = Attachment.getEncodingFromString(encoding);
+        this.length = length;
+        this.encodedLength = encodedLength;
+    }
+
+    @Override
+    public PreparedAttachment processStream(InputStream stream) throws AttachmentException {
+        UnsavedStreamAttachment usa = new UnsavedStreamAttachment(stream, contentType, encoding);
+        PreparedAttachment attachment = datastoreWrapper.prepareAttachment(usa, length,
+                encodedLength);
+        return attachment;
+    }
+}

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/CouchClientWrapper.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/CouchClientWrapper.java
@@ -104,10 +104,25 @@ public class CouchClientWrapper implements CouchDB {
         Misc.checkNotNullOrEmpty(checkpointId, "Checkpoint id");
         Misc.checkNotNullOrEmpty(sequence, "Sequence");
         String replicatorLocalDocId = getCheckpointLocalDocId(checkpointId);
-        if (couchClient.contains(replicatorLocalDocId)) {
-            updateCheckpoint(replicatorLocalDocId, sequence);
-        } else {
-            createCheckpoint(replicatorLocalDocId, sequence);
+
+        RemoteCheckpointDoc checkpointDoc;
+        try {
+            checkpointDoc = couchClient.getDocument(replicatorLocalDocId, RemoteCheckpointDoc
+                    .class);
+            // If it existed, update the sequence
+            checkpointDoc.setLastSequence(sequence);
+        } catch (NoResourceException e) {
+            // Didn't exist yet, but we'll create it now so don't worry
+            checkpointDoc = new RemoteCheckpointDoc(sequence);
+        }
+
+        // Write the new or updated doc
+        try {
+            Response response = couchClient.putUpdate(replicatorLocalDocId, checkpointDoc);
+            logger.fine(String.format("Response: %s", response));
+        } catch (DocumentConflictException e) {
+            // Could happen if we retried, but had already succeeded, so suppress.
+            // Worst case the checkpoint isn't written and we replicate a little extra next time.
         }
     }
 
@@ -190,22 +205,6 @@ public class CouchClientWrapper implements CouchDB {
     @Override
     public Response delete(String id, String rev) {
         return couchClient.delete(id, rev);
-    }
-
-
-    private void createCheckpoint(String checkpointDocId, String sequence) {
-        RemoteCheckpointDoc checkpointDoc = new RemoteCheckpointDoc(sequence);
-        checkpointDoc.setId(checkpointDocId);
-        Response response = couchClient.create(checkpointDoc);
-        logger.fine(String.format("Response: %s",response));
-    }
-
-    private void updateCheckpoint(String checkpointDocId, String sequence) {
-        RemoteCheckpointDoc checkpointDoc = couchClient.getDocument(
-                checkpointDocId, RemoteCheckpointDoc.class);
-        checkpointDoc.setLastSequence(sequence);
-        Response response = couchClient.update(checkpointDocId, checkpointDoc);
-        logger.fine(String.format("Response: %s",response));
     }
 
     public void createDatabase() {

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/CouchClientWrapper.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/CouchClientWrapper.java
@@ -1,7 +1,7 @@
 /*
- * Copyright © 2013 Cloudant, Inc. All rights reserved.
+ * Copyright © 2016, 2017 IBM Corp. All rights reserved.
  *
- * Copyright © 2016 IBM Corp. All rights reserved.
+ * Copyright © 2013 Cloudant, Inc. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
  * except in compliance with the License. You may obtain a copy of the License at
@@ -18,22 +18,21 @@ package com.cloudant.sync.internal.replication;
 
 import com.cloudant.http.HttpConnectionRequestInterceptor;
 import com.cloudant.http.HttpConnectionResponseInterceptor;
+import com.cloudant.sync.internal.documentstore.DocumentRevsList;
+import com.cloudant.sync.internal.documentstore.InternalDocumentRevision;
+import com.cloudant.sync.internal.documentstore.MultipartAttachmentWriter;
 import com.cloudant.sync.internal.mazha.ChangesResult;
 import com.cloudant.sync.internal.mazha.CouchClient;
 import com.cloudant.sync.internal.mazha.CouchException;
+import com.cloudant.sync.internal.mazha.DocumentConflictException;
 import com.cloudant.sync.internal.mazha.DocumentRevs;
+import com.cloudant.sync.internal.mazha.NoResourceException;
 import com.cloudant.sync.internal.mazha.OkOpenRevision;
 import com.cloudant.sync.internal.mazha.OpenRevision;
 import com.cloudant.sync.internal.mazha.Response;
-import com.cloudant.sync.documentstore.Attachment;
-import com.cloudant.sync.internal.documentstore.InternalDocumentRevision;
-import com.cloudant.sync.internal.documentstore.DocumentRevsList;
-import com.cloudant.sync.internal.documentstore.MultipartAttachmentWriter;
-import com.cloudant.sync.documentstore.UnsavedStreamAttachment;
-import com.cloudant.sync.replication.PullFilter;
 import com.cloudant.sync.internal.util.Misc;
+import com.cloudant.sync.replication.PullFilter;
 
-import java.io.InputStream;
 import java.net.URI;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -265,17 +264,14 @@ public class CouchClientWrapper implements CouchDB {
     }
 
     @Override
-    public UnsavedStreamAttachment getAttachmentStream(String id, String rev, String attachmentName, String contentType, String encodingStr) {
-        // call with last arg `true` to indicate we will accept gzipped data
-        InputStream is = this.couchClient.getAttachmentStream(id, rev, attachmentName, true);
-        Attachment.Encoding encoding = Attachment.getEncodingFromString(encodingStr);
-        UnsavedStreamAttachment usa = new UnsavedStreamAttachment(is, contentType, encoding);
-        return usa;
+    public boolean isBulkSupported() {
+        return this.couchClient.isBulkSupported();
     }
 
     @Override
-    public boolean isBulkSupported() {
-        return this.couchClient.isBulkSupported();
+    public <T> T pullAttachmentWithRetry(String id, String rev, String name, CouchClient
+            .InputStreamProcessor<T> streamProcessor) {
+        return couchClient.processAttachmentStream(id, rev, name, true, streamProcessor);
     }
 
 }

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/CouchClientWrapper.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/CouchClientWrapper.java
@@ -241,6 +241,8 @@ public class CouchClientWrapper implements CouchDB {
         }
 
         List<Response> responses = couchClient.bulkCreateSerializedDocs(serializedDocs);
+        // N.B. The successful response to a _bulk_docs using new_edits:false is an empty array
+        // hence this unusual check and throw exception if any responses are returned.
         if(responses != null && responses.size() > 0) {
             logger.severe(String.format("Unknown bulkCreateDocs API error: %s for input: %s",responses,serializedDocs));
             throw new RuntimeException("Unknown bulkCreateDocs api error");

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/CouchDB.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/CouchDB.java
@@ -16,14 +16,13 @@
 
 package com.cloudant.sync.internal.replication;
 
+import com.cloudant.sync.internal.documentstore.InternalDocumentRevision;
+import com.cloudant.sync.internal.documentstore.DocumentRevsList;
+import com.cloudant.sync.internal.documentstore.MultipartAttachmentWriter;
 import com.cloudant.sync.internal.mazha.ChangesResult;
 import com.cloudant.sync.internal.mazha.CouchClient;
 import com.cloudant.sync.internal.mazha.DocumentRevs;
 import com.cloudant.sync.internal.mazha.Response;
-import com.cloudant.sync.internal.documentstore.InternalDocumentRevision;
-import com.cloudant.sync.internal.documentstore.DocumentRevsList;
-import com.cloudant.sync.internal.documentstore.MultipartAttachmentWriter;
-import com.cloudant.sync.documentstore.UnsavedStreamAttachment;
 import com.cloudant.sync.replication.PullFilter;
 
 import java.util.Collection;
@@ -58,10 +57,11 @@ interface CouchDB {
     void bulkCreateSerializedDocs(List<String> serializedDocs);
     List<Response> putMultiparts(List<MultipartAttachmentWriter> multiparts);
     Map<String, CouchClient.MissingRevisions> revsDiff(Map<String, Set<String>> revisions);
-    UnsavedStreamAttachment getAttachmentStream(String id, String rev, String attachmentName, String contentType, String encoding);
 
     Iterable<DocumentRevsList> bulkGetRevisions(List<BulkGetRequest> requests,
                                                    boolean pullAttachmentsInline);
     boolean isBulkSupported();
+
+    <T> T pullAttachmentWithRetry(String id, String rev, String name, CouchClient.InputStreamProcessor<T> streamProcessor);
 
 }

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/PullStrategy.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/PullStrategy.java
@@ -18,23 +18,23 @@ package com.cloudant.sync.internal.replication;
 
 import com.cloudant.http.HttpConnectionRequestInterceptor;
 import com.cloudant.http.HttpConnectionResponseInterceptor;
-import com.cloudant.sync.internal.mazha.ChangesResult;
-import com.cloudant.sync.internal.mazha.CouchClient;
-import com.cloudant.sync.internal.mazha.DocumentRevs;
 import com.cloudant.sync.documentstore.Attachment;
 import com.cloudant.sync.documentstore.Database;
 import com.cloudant.sync.internal.documentstore.DatabaseImpl;
 import com.cloudant.sync.documentstore.DocumentStoreException;
 import com.cloudant.sync.documentstore.DocumentException;
+import com.cloudant.sync.event.EventBus;
 import com.cloudant.sync.internal.documentstore.DocumentRevsList;
 import com.cloudant.sync.internal.documentstore.PreparedAttachment;
-import com.cloudant.sync.documentstore.UnsavedStreamAttachment;
-import com.cloudant.sync.event.EventBus;
-import com.cloudant.sync.replication.DatabaseNotFoundException;
-import com.cloudant.sync.replication.PullFilter;
+import com.cloudant.sync.internal.mazha.ChangesResult;
+import com.cloudant.sync.internal.mazha.CouchClient;
+import com.cloudant.sync.internal.mazha.DocumentRevs;
 import com.cloudant.sync.internal.util.CollectionUtils;
 import com.cloudant.sync.internal.util.JSONUtils;
 import com.cloudant.sync.internal.util.Misc;
+import com.cloudant.sync.replication.DatabaseNotFoundException;
+import com.cloudant.sync.replication.PullFilter;
+
 
 import org.apache.commons.codec.binary.Hex;
 
@@ -382,15 +382,14 @@ public class PullStrategy implements ReplicationStrategy {
                                         }
 
                                     }
-                                    UnsavedStreamAttachment usa = this.sourceDb
-                                            .getAttachmentStream(documentRevs.getId(),
-                                                    documentRevs.getRev(), attachmentName,
-                                                    contentType, encoding);
 
                                     // by preparing the attachment here, it is downloaded outside
                                     // of the database transaction
-                                    preparedAtts.put(attachmentName, this.targetDb.prepareAttachment(usa, length,
-                                            encodedLength));
+                                    preparedAtts.put(attachmentName, this.sourceDb.pullAttachmentWithRetry
+                                            (documentRevs.getId(), documentRevs.getRev(), entry
+                                                    .getKey(), new AttachmentPullProcessor(this
+                                                    .targetDb, entry.getKey(), contentType,
+                                                    encoding, length, encodedLength)));
                                 }
                             }
                         } catch (Exception e) {

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/util/JSONUtils.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/util/JSONUtils.java
@@ -20,7 +20,6 @@ import com.cloudant.sync.internal.common.CouchConstants;
 import com.cloudant.sync.internal.common.PropertyFilterMixIn;
 import com.cloudant.sync.internal.mazha.CouchClient;
 import com.cloudant.sync.internal.mazha.Document;
-import com.cloudant.sync.internal.mazha.DocumentRevs;
 import com.cloudant.sync.internal.mazha.OpenRevision;
 import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.core.JsonProcessingException;
@@ -32,7 +31,6 @@ import com.fasterxml.jackson.databind.ObjectWriter;
 import com.fasterxml.jackson.databind.ser.FilterProvider;
 import com.fasterxml.jackson.databind.ser.impl.SimpleBeanPropertyFilter;
 import com.fasterxml.jackson.databind.ser.impl.SimpleFilterProvider;
-import com.fasterxml.jackson.databind.type.TypeFactory;
 
 import java.io.IOException;
 import java.io.Reader;
@@ -52,13 +50,18 @@ public class JSONUtils {
 
     private static final byte[] EMPTY_JSON = "{}".getBytes(Charset.forName("UTF-8"));
 
-
     public final static TypeReference<List<String>> STRING_LIST_TYPE_DEF =
             new TypeReference<List<String>>() {};
 
     public final static TypeReference<Map<String, Object>> STRING_MAP_TYPE_DEF =
             new TypeReference<Map<String, Object>>() {};
 
+    public final static TypeReference<Map<String, CouchClient
+            .MissingRevisions>> STRING_MISSING_REVS_MAP_TYPE_DEF =
+            new TypeReference<Map<String, CouchClient.MissingRevisions>>() {};
+
+    public final static TypeReference<List<OpenRevision>> OPEN_REVS_LIST_TYPE_DEF =
+            new TypeReference<List<OpenRevision>>() {};
 
     private static final ObjectMapper sMapper = new ObjectMapper();
 
@@ -237,26 +240,5 @@ public class JSONUtils {
         } catch (JsonProcessingException e) {
             throw new RuntimeException(e);
         }
-    }
-
-    public static TypeFactory getTypeFactory(){
-        return sMapper.getTypeFactory();
-    }
-
-    public static JavaType mapStringToObject(){
-        return JSONUtils.getTypeFactory().constructParametricType(Map.class,String.class,Object.class);
-    }
-
-    public static JavaType mapStringMissingRevisions(){
-        return JSONUtils.getTypeFactory().constructParametricType(Map.class, String.class, CouchClient
-                .MissingRevisions.class);
-    }
-
-    public static JavaType openRevisionList() {
-        return JSONUtils.getTypeFactory().constructParametricType(List.class, OpenRevision.class);
-    }
-
-    public static JavaType documentRevs() {
-        return JSONUtils.getTypeFactory().constructType(DocumentRevs.class);
     }
 }

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/mazha/BulkAPITest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/mazha/BulkAPITest.java
@@ -16,9 +16,11 @@
 
 package com.cloudant.sync.internal.mazha;
 
+import static org.hamcrest.CoreMatchers.hasItem;
+
+import com.cloudant.common.RequireRunningCouchDB;
 import com.cloudant.sync.internal.common.CouchConstants;
 import com.cloudant.sync.internal.common.CouchUtils;
-import com.cloudant.common.RequireRunningCouchDB;
 import com.cloudant.sync.internal.util.JSONUtils;
 import com.cloudant.sync.util.TestUtils;
 
@@ -32,8 +34,6 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
-
-import static org.hamcrest.CoreMatchers.hasItem;
 
 @Category(RequireRunningCouchDB.class)
 public class BulkAPITest extends CouchClientTestBase {
@@ -82,7 +82,7 @@ public class BulkAPITest extends CouchClientTestBase {
         List<Response> responses = client.bulkCreateDocs(doc1);
         Assert.assertEquals(0, responses.size());
 
-        Map<String, Object> allRevs = client.getDocRevisions(res1.getId(), revision3, JSONUtils.mapStringToObject());
+        Map<String, Object> allRevs = client.getDocRevisions(res1.getId(), revision3, JSONUtils.STRING_MAP_TYPE_DEF);
 
         int revisionStart = (Integer) ((Map<String, Object>) allRevs.get(CouchConstants._revisions)).get
                 (CouchConstants.start);

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/mazha/CouchClientBasicTest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/mazha/CouchClientBasicTest.java
@@ -26,15 +26,12 @@ import com.cloudant.common.RequireRunningCouchDB;
 import com.cloudant.sync.internal.util.JSONUtils;
 import com.cloudant.sync.internal.util.Misc;
 
-import org.apache.commons.io.IOUtils;
 import org.junit.Assert;
 import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 
 import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
@@ -171,18 +168,6 @@ public class CouchClientBasicTest extends CouchClientTestBase {
     }
 
     @Test
-    public void getDocumentInputStream_id_success() {
-        Response res = ClientTestUtils.createHelloWorldDoc(client);
-        InputStream is = client.getDocumentStream(res.getId(), res.getRev());
-        try {
-            Map<String, Object> doc = JSONUtils.fromJson(new InputStreamReader(is));
-            assertHelloWorldMapObject(res, doc);
-        } finally {
-            IOUtils.closeQuietly(is);
-        }
-    }
-
-    @Test
     public void getDocument_id_success() {
         Response res = ClientTestUtils.createHelloWorldDoc(client);
         Map<String, Object> doc = client.getDocument(res.getId());
@@ -196,11 +181,6 @@ public class CouchClientBasicTest extends CouchClientTestBase {
         Assert.assertEquals(res.getId(), doc.get(CouchConstants._id));
         Assert.assertTrue(doc.containsKey(CouchConstants._rev));
         Assert.assertEquals(res.getRev(), doc.get(CouchConstants._rev));
-    }
-
-    @Test(expected = NoResourceException.class)
-    public void getDocumentInputStream_idNotExist_exception() {
-        client.getDocumentStream("id_not_exist", "1-no_such_rev");
     }
 
     @Test(expected = NoResourceException.class)
@@ -245,24 +225,6 @@ public class CouchClientBasicTest extends CouchClientTestBase {
             Assert.assertNotNull(fooVersion1);
             Assert.assertThat(fooVersion1.getRevision(), startsWith("1-"));
         }
-    }
-
-    @Test
-    public void getDocumentInputStream_idRev_success() {
-        Response res = ClientTestUtils.createHelloWorldDoc(client);
-        InputStream is = client.getDocumentStream(res.getId(), res.getRev());
-        try {
-            Map<String, Object> doc = JSONUtils.fromJson(new InputStreamReader(is));
-            assertHelloWorldMapObject(res, doc);
-        } finally {
-            IOUtils.closeQuietly(is);
-        }
-    }
-
-    @Test(expected = NoResourceException.class)
-    public void getDocumentInputStream_idRevNotExist_exception() {
-        Response res = ClientTestUtils.createHelloWorldDoc(client);
-        client.getDocumentStream(res.getId(), "1-revnotexist");
     }
 
     @Test(expected = NoResourceException.class)

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/mazha/CouchClientBasicTest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/mazha/CouchClientBasicTest.java
@@ -410,4 +410,11 @@ public class CouchClientBasicTest extends CouchClientTestBase {
             return false;
         }
     }
+
+    @Test
+    public void getRevision() {
+        Response res = ClientTestUtils.createHelloWorldDoc(client);
+        String rev = client.getDocumentRev(res.getId());
+        Assert.assertEquals("The revisions should be the same", res.getRev(), rev);
+    }
 }

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/mazha/GetDocumentRevsTest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/mazha/GetDocumentRevsTest.java
@@ -16,9 +16,16 @@
 
 package com.cloudant.sync.internal.mazha;
 
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.hasItem;
+import static org.hamcrest.CoreMatchers.hasItems;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.CoreMatchers.startsWith;
+import static org.hamcrest.core.Is.is;
+
+import com.cloudant.common.RequireRunningCouchDB;
 import com.cloudant.sync.internal.common.CouchConstants;
 import com.cloudant.sync.internal.common.CouchUtils;
-import com.cloudant.common.RequireRunningCouchDB;
 import com.cloudant.sync.internal.util.JSONUtils;
 
 import org.junit.Assert;
@@ -27,9 +34,6 @@ import org.junit.experimental.categories.Category;
 
 import java.util.List;
 import java.util.Map;
-
-import static org.hamcrest.CoreMatchers.*;
-import static org.hamcrest.core.Is.is;
 
 @Category(RequireRunningCouchDB.class)
 public class GetDocumentRevsTest extends CouchClientTestBase {
@@ -53,7 +57,7 @@ public class GetDocumentRevsTest extends CouchClientTestBase {
     public void getDocRevisions_idAndRevAndMapClass_revsMustBeReturned() {
         Response[] responses = createAndUpdateDocumentAndReturnRevisions();
         Map<String, Object> revs = client.getDocRevisions(responses[1].getId(), responses[1].getRev(),
-                JSONUtils.mapStringToObject());
+                JSONUtils.STRING_MAP_TYPE_DEF);
 
         Assert.assertThat(revs.keySet(), hasItem(CouchConstants._revisions));
 

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/AttachmentsPullTest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/AttachmentsPullTest.java
@@ -17,8 +17,8 @@
 package com.cloudant.sync.internal.replication;
 
 import com.cloudant.common.RequireRunningCouchDB;
-import com.cloudant.sync.internal.mazha.Response;
 import com.cloudant.sync.documentstore.Attachment;
+import com.cloudant.sync.internal.mazha.Response;
 import com.cloudant.sync.util.TestUtils;
 
 import org.junit.Assert;
@@ -33,6 +33,7 @@ import java.io.FileInputStream;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.concurrent.TimeUnit;
 
 /**
  * Created by tomblench on 26/03/2014.
@@ -180,8 +181,6 @@ public class AttachmentsPullTest extends ReplicationTestBase {
         bar._attachments = null;
 
         Response res = remoteDb.update(id, bar);
-        bar = remoteDb.get(BarWithAttachments.class, res.getId());
-
         rev = res.getRev();
     }
 
@@ -192,18 +191,15 @@ public class AttachmentsPullTest extends ReplicationTestBase {
         bar.setAge(35);
 
         Response res = remoteDb.update(id, bar);
-        bar = remoteDb.get(BarWithAttachments.class, res.getId());
-
         rev = res.getRev();
     }
 
-    private void createRevisionAndAttachment() {
+    private void createRevisionAndAttachment() throws Exception {
         BarWithAttachments bar = new BarWithAttachments();
         bar.setName("Tom");
         bar.setAge(34);
 
         Response res = remoteDb.create(bar);
-        bar = remoteDb.get(BarWithAttachments.class, res.getId());
 
         id = res.getId();
         rev = res.getRev();
@@ -211,11 +207,10 @@ public class AttachmentsPullTest extends ReplicationTestBase {
                 attachmentData.getBytes());
 
         // putting attachment will have updated the rev
-        bar = remoteDb.get(BarWithAttachments.class, res.getId());
-        rev = bar.getRevision();
+        rev = getLatestRevision(id, rev);
     }
 
-    private void createRevisionAndBigAttachment() throws IOException {
+    private void createRevisionAndBigAttachment() throws Exception {
         Bar bar = new Bar();
         bar.setName("Tom");
         bar.setAge(34);
@@ -234,11 +229,10 @@ public class AttachmentsPullTest extends ReplicationTestBase {
                 data);
 
         // putting attachment will have updated the rev
-        bar = remoteDb.get(Bar.class, res.getId());
-        rev = bar.getRevision();
+        rev = getLatestRevision(id, rev);
     }
 
-    private void createRevisionAndBigTextAttachment() throws IOException {
+    private void createRevisionAndBigTextAttachment() throws Exception {
         Bar bar = new Bar();
         bar.setName("Tom");
         bar.setAge(34);
@@ -257,12 +251,11 @@ public class AttachmentsPullTest extends ReplicationTestBase {
                 "text/plain", data);
 
         // putting attachment will have updated the rev
-        bar = remoteDb.get(Bar.class, res.getId());
-        rev = bar.getRevision();
+        rev = getLatestRevision(id, rev);
     }
 
 
-    private void updateRevisionAndAttachment() {
+    private void updateRevisionAndAttachment() throws Exception {
         Bar bar = new Bar();
         bar.setName("Dick");
         bar.setAge(33);
@@ -274,8 +267,26 @@ public class AttachmentsPullTest extends ReplicationTestBase {
                 attachmentData2.getBytes());
 
         // putting attachment will have updated the rev
-        bar = remoteDb.get(Bar.class, res.getId());
-        rev = bar.getRevision();
+        rev = getLatestRevision(id, rev);
+    }
+
+    /**
+     * Utility for reading the revision of a document modified by an attachment write.
+     * When we are writing attachments it might take a little time for new revision to appear so we
+     * validate that the rev has incremented from what we currently know about before returning the
+     * latest rev.
+     *
+     * @param docId      the ID of the document to read
+     * @param currentRev the revision currently known about
+     * @return a newer revision identifier
+     */
+    private String getLatestRevision(String docId, String currentRev) throws Exception {
+        String newRev = null;
+        do {
+            newRev = remoteDb.getCouchClient().getDocumentRev(docId);
+            TimeUnit.MILLISECONDS.sleep(500l);
+        } while (currentRev.equals(newRev));
+        return newRev;
     }
 
     // override so we can have custom value for pullAttachmentsInline

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/AttachmentsPushTest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/AttachmentsPushTest.java
@@ -19,6 +19,7 @@ import com.cloudant.sync.documentstore.Attachment;
 import com.cloudant.sync.documentstore.DocumentBodyFactory;
 import com.cloudant.sync.documentstore.DocumentRevision;
 import com.cloudant.sync.documentstore.UnsavedFileAttachment;
+import com.cloudant.sync.internal.mazha.CouchClient;
 import com.cloudant.sync.replication.PushAttachmentsInline;
 import com.cloudant.sync.replication.ReplicatorBuilder;
 import com.cloudant.sync.util.TestUtils;
@@ -31,13 +32,12 @@ import org.junit.runners.Parameterized;
 
 import java.io.File;
 import java.io.FileInputStream;
+import java.io.FileNotFoundException;
 import java.io.InputStream;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
-
-import static org.hamcrest.CoreMatchers.is;
 
 /**
  * Created by tomblench on 26/03/2014.
@@ -133,9 +133,7 @@ public class AttachmentsPushTest extends ReplicationTestBase {
         push();
 
         // check it's in the DB
-        InputStream is1 = this.couchClient.getAttachmentStream(id1, newRevision.getRevision(), attachmentName, false);
-        InputStream is2 = new FileInputStream(f);
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(is1, is2));
+        assertAttachmentEquality(id1, newRevision.getRevision(), attachmentName, f);
     }
 
     @Test
@@ -157,9 +155,7 @@ public class AttachmentsPushTest extends ReplicationTestBase {
         push();
 
         // check it's in the DB
-        InputStream is1 = this.couchClient.getAttachmentStream(id1, newRevision.getRevision(), attachmentName, false);
-        InputStream is2 = new FileInputStream(f);
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(is1, is2));
+        assertAttachmentEquality(id1, newRevision.getRevision(), attachmentName, f);
     }
 
 
@@ -188,12 +184,8 @@ public class AttachmentsPushTest extends ReplicationTestBase {
         push();
 
         // check it's in the DB
-        InputStream is1_1 = this.couchClient.getAttachmentStream(id1, newRevision.getRevision(), attachmentName1, false);
-        InputStream is1_2 = this.couchClient.getAttachmentStream(id1, newRevision.getRevision(), attachmentName2, false);
-        InputStream is2_1 = new FileInputStream(f1);
-        InputStream is2_2 = new FileInputStream(f2);
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(is1_1, is2_1));
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(is1_2, is2_2));
+        assertAttachmentEquality(id1, newRevision.getRevision(), attachmentName1, f1);
+        assertAttachmentEquality(id1, newRevision.getRevision(), attachmentName2, f2);
     }
 
     @Test
@@ -231,25 +223,10 @@ public class AttachmentsPushTest extends ReplicationTestBase {
         // push replication - att2 should be uploaded
         push();
 
-        InputStream isOriginal1;
-        InputStream isOriginal2;
-
-        InputStream isRev2 = this.couchClient.getAttachmentStream(id1, rev2.getRevision(), attachmentName1, false);
-        InputStream isRev3 = this.couchClient.getAttachmentStream(id1, rev3.getRevision(), attachmentName1, false);
-        InputStream isRev4Att1 = this.couchClient.getAttachmentStream(id1, rev4.getRevision(), attachmentName1, false);
-        InputStream isRev4Att2 = this.couchClient.getAttachmentStream(id1, rev4.getRevision(), attachmentName2, false);
-
-        isOriginal1 = new FileInputStream(f1);
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(isRev2, isOriginal1));
-
-        isOriginal1 = new FileInputStream(f1);
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(isRev3, isOriginal1));
-
-        isOriginal1 = new FileInputStream(f1);
-        isOriginal2 = new FileInputStream(f2);
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(isRev4Att1, isOriginal1));
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(isRev4Att2, isOriginal2));
-
+        assertAttachmentEquality(id1, rev2.getRevision(), attachmentName1, f1);
+        assertAttachmentEquality(id1, rev3.getRevision(), attachmentName1, f1);
+        assertAttachmentEquality(id1, rev4.getRevision(), attachmentName1, f1);
+        assertAttachmentEquality(id1, rev4.getRevision(), attachmentName2, f2);
     }
 
     // regression test for FB 46326 - see
@@ -283,17 +260,27 @@ public class AttachmentsPushTest extends ReplicationTestBase {
         // push replication - att1 & att2 should be uploaded
         push();
 
-        InputStream isOriginal1;
-        InputStream isOriginal2;
+        assertAttachmentEquality(id1, rev4.getRevision(), attachmentName1, f1);
+        assertAttachmentEquality(id1, rev4.getRevision(), attachmentName2, f2);
+    }
 
-        InputStream isRev4Att1 = this.couchClient.getAttachmentStream(id1, rev4.getRevision(), attachmentName1, false);
-        InputStream isRev4Att2 = this.couchClient.getAttachmentStream(id1, rev4.getRevision(), attachmentName2, false);
+    void assertAttachmentEquality(String id, String rev, String attachmentName, File f) throws Exception {
+        this.couchClient.processAttachmentStream(id, rev, attachmentName, false, new FileComparingInputStreamProcessor(f));
+    }
 
-        isOriginal1 = new FileInputStream(f1);
-        isOriginal2 = new FileInputStream(f2);
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(isRev4Att1, isOriginal1));
-        Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(isRev4Att2, isOriginal2));
+    private static final class FileComparingInputStreamProcessor implements CouchClient.InputStreamProcessor<Void> {
 
+        private final FileInputStream fileStream;
+
+        FileComparingInputStreamProcessor(File f) throws FileNotFoundException {
+            this.fileStream = new FileInputStream(f);
+        }
+
+        @Override
+        public Void processStream(InputStream stream) throws Exception {
+            Assert.assertTrue("Attachment not the same", TestUtils.streamsEqual(stream, fileStream));
+            return null;
+        }
     }
 
 }

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/DBWithSlashReplicationTest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/DBWithSlashReplicationTest.java
@@ -155,7 +155,7 @@ public class DBWithSlashReplicationTest extends ReplicationTestBase {
     }
 
     @Override
-    String getDbName() {
-        return "dbwith%2Faslash";
+    protected String getDbName() {
+        return "dbwith%2Faslash_" + super.getDbName();
     }
 }

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/PullReplicationTerminationTest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/PullReplicationTerminationTest.java
@@ -83,7 +83,7 @@ public class PullReplicationTerminationTest extends ReplicationTestBase {
     @Before
     public void customizeReplicatorAndPopulateDb() throws Exception {
         setupTerminationTestReplicator();
-        populateRemoteDb(DOC_BATCHES, DOCS_PER_BATCH);
+        populateRemoteDb();
     }
 
     private void setupTerminationTestReplicator() {
@@ -93,7 +93,7 @@ public class PullReplicationTerminationTest extends ReplicationTestBase {
     }
 
 
-    private void populateRemoteDb(int batches, int docsPerBatch) throws Exception {
+    private void populateRemoteDb() throws Exception {
         // Create documents in the remote db in batches
         for (List<Foo> batch : GENERATED_FOO_DOC_BATCHES) {
             remoteDb.getCouchClient().bulkCreateDocs(batch);

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/ReplicationTestBase.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/ReplicationTestBase.java
@@ -20,14 +20,15 @@ import com.cloudant.common.CouchTestBase;
 import com.cloudant.http.HttpConnectionRequestInterceptor;
 import com.cloudant.http.HttpConnectionResponseInterceptor;
 import com.cloudant.http.interceptors.CookieInterceptor;
+import com.cloudant.sync.documentstore.DocumentStore;
 import com.cloudant.sync.internal.mazha.CouchClient;
 import com.cloudant.sync.internal.mazha.CouchConfig;
+import com.cloudant.sync.internal.mazha.CouchException;
+import com.cloudant.sync.internal.sqlite.SQLDatabase;
 import com.cloudant.sync.internal.documentstore.DatabaseImpl;
-import com.cloudant.sync.documentstore.DocumentStore;
 import com.cloudant.sync.replication.PullFilter;
 import com.cloudant.sync.replication.Replicator;
 import com.cloudant.sync.replication.ReplicatorBuilder;
-import com.cloudant.sync.internal.sqlite.SQLDatabase;
 import com.cloudant.sync.util.TestUtils;
 
 import org.junit.After;
@@ -113,7 +114,13 @@ public abstract class ReplicationTestBase extends CouchTestBase {
                 couchConfig.getRootUri(),
                 requestInterceptors,
                 responseInterceptors));
-        remoteDb.createDatabase();
+        try {
+            remoteDb.createDatabase();
+        } catch (CouchException e) {
+            // Suppress 412 exceptions in the event a retry has caused us to create twice, but
+            // usually throw
+            if (e.getStatusCode() != 412) throw e;
+        }
         couchClient = remoteDb.getCouchClient();
     }
 
@@ -122,7 +129,7 @@ public abstract class ReplicationTestBase extends CouchTestBase {
         CouchClientWrapperDbUtils.deleteDbQuietly(remoteDb);
     }
 
-    String getDbName() {
+    protected String getDbName() {
         String dbName = getClass().getSimpleName()+ dbSuffix;
         String regex = "([a-z])([A-Z])";
         String replacement = "$1_$2";

